### ISSUE TITLE
Query with case condition

### ISSUE THREAD

**Issue by tomtalp** [tgriesser/knex#1482]
Hey, couldn't find an answer for this in the docs or anywhere online - is there a way to build a query with conditions? Something such as -
```
SELECT CASE WHEN a = 1
       THEN 1
       ELSE 0 
    END as test123
```
?

And if that's possible, can something a bit more complex like this be done? 
```
SUM sum(CASE WHEN a = 1
    THEN 1
      ELSE 0 END) AS test123
```

Thanks

**Comment by rhys-vdw** [comment]
Yes, this is possible, but not nearly as nice as one might hope:

```js
knex.raw(`SELECT CASE WHEN a = 1
  THEN 1
  ELSE 0 
  END as test123
`).then(result =>
```

Or more likely what you want:

```js
const sumA = knex.raw(`sum(SELECT CASE WHEN a = 1
  THEN 1
  ELSE 0 
  END
) as test123`);

knex('some_table').select(sumA)
```

NOTES:

*Unable to chain `.as` onto the end of a `raw` query.*

```js
// doesn't work:
knex.raw('select 1').as('alias');
```

*Unable to pass raw to `sum`*

```js
// doesn't work
knex('table').sum(knex.raw('select 1'))
```

*Expected API*

```js
// doesn't work
const sumA = knex.raw(`SELECT CASE WHEN a = 1
  THEN 1
  ELSE 0 
  END
`);

knex('some_table').sum(sumA.as('test123'))
```

**Comment by rhys-vdw** [comment]
Added 'PR Please' for the various changes I've listed in the above post under 'NOTES:' if anybody is interested.

**Comment by ericadamski** [comment]
If there is still a need for this type of functionality, here is an implementation that I would be interested in PR'ing here. https://www.npmjs.com/package/knex-case

Let me know if anyone has any opinions on this!

**Comment by elhigu** [comment]
Sounds kind of interesting. There has been also some other needs for building other kind of expressions which are just part of the query. This would fit nicely to that kind of expression builder.

**Comment by serv** [comment]
This worked for me

```javascript
const conditionalCount = knex.raw(`
  CASE WHEN some_column=true 
  THEN some_column_to_sum
  ELSE 0
  END
`);

knex('table_name').sum({ aliased_sum: conditionalCount });
```

**Comment by mercmobily** [comment]
Thank you for reporting this issue and providing clear examples of the limitations with CASE WHEN expressions in Knex. You've identified several important gaps in the API that would make complex aggregations much more ergonomic to work with.

This issue remains open and relevant, as the current workarounds are not ideal. The community has provided some solutions, but none match the clean API you've proposed. Let me address each point you raised.

The first workaround you showed, using raw queries directly, does work but lacks the composability and type safety that makes Knex valuable. Having to write the entire SQL expression as a string defeats much of the purpose of using a query builder.

```javascript
knex.raw(`SELECT CASE WHEN a = 1
  THEN 1
  ELSE 0 
  END as test123
`)
```

The second approach, wrapping the CASE statement in the aggregate function, is closer to what users expect but still requires writing raw SQL:

```javascript
const sumA = knex.raw(`sum(SELECT CASE WHEN a = 1
  THEN 1
  ELSE 0 
  END
) as test123`);

knex('some_table').select(sumA)
```

Looking at the codebase, I can confirm that the limitations you identified are still present in the current version. The Raw class in lib/raw.js:23-195 does not implement an `.as()` method, which prevents the natural chaining pattern you've described. The aggregate functions in lib/query/querycompiler.js:331-410 also don't support Raw objects as direct arguments, requiring string values instead.

The community has found a workaround that's slightly better, as mentioned by @serv in the comments. You can pass a Raw object to the aggregate function using an object syntax:

```javascript
const conditionalCount = knex.raw(`
  CASE WHEN some_column=true 
  THEN some_column_to_sum
  ELSE 0
  END
`);

knex('table_name').sum({ aliased_sum: conditionalCount });
```

This approach works because the object form bypasses some of the type checking, but it's still not as clean as the proposed API would be.

There was an attempt to address this through the knex-case package that @ericadamski mentioned, which provides a more structured way to build CASE statements. However, that package hasn't been updated in 7 years and appears to be abandoned, indicating this is still an unmet need in the ecosystem.

The maintainer @elhigu expressed interest in this functionality, noting it would fit well into a broader expression builder pattern. This suggests the feature would be welcomed if properly implemented.

For anyone wanting to contribute a PR to address this, the key changes needed would be:

1. Adding an `.as()` method to the Raw class that wraps the raw SQL with an alias
2. Modifying the aggregate functions to accept Raw objects directly
3. Ensuring proper handling of Raw objects throughout the query compilation process

Until this is implemented in core, the best current approach is to use the object syntax with Raw queries for aggregate functions, or to use Raw queries in select statements when you need full control over the SQL.

Related issues:
- #238 - Early discussion about aggregate functions with complex expressions
- #1606 - How to perform complex Sums (closed with workaround)
- #2823 - Request for knex.val() similar to knex.ref() which could help with expression building
